nav2_route/src/path_converter.cpp

```diff
@@ -54,17 +61,44 @@ nav_msgs::msg::Path PathConverter::densify(
     interpolateEdge(start.x, start.y, end.x, end.y, path.poses);
   }
 
+  Coordinates start = route.edges[0]->start->coords;
+  Coordinates end;
+
   // Fill in path via route edges
-  for (unsigned int i = 0; i != route.edges.size(); i++) {
+
+  for (unsigned int i = 0; i < route.edges.size() - 1; i++) {
     const EdgePtr edge = route.edges[i];
-    const Coordinates & start = edge->start->coords;
-    const Coordinates & end = edge->end->coords;
-    interpolateEdge(start.x, start.y, end.x, end.y, path.poses);
+    const EdgePtr next_edge = route.edges[i + 1];
+
+    end = edge->end->coords;
+
+    CornerArc corner_arc(start, end, next_edge->end->coords, smoothing_radius_);
+
+    if (corner_arc.isCornerValid() && smooth_corners_) {
+      // if an arc exists, end of the first edge is the start of the arc
+      end = corner_arc.getCornerStart();
+
+      // interpolate to start of arc
+      interpolateEdge(start.x, start.y, end.x, end.y, path.poses);
+
+      // interpolate arc
+      corner_arc.interpolateArc(density_ / smoothing_radius_, path.poses);
+
+      // new start of next edge is end of smoothing arc
+      start = corner_arc.getCornerEnd();
+    } else {
+      interpolateEdge(start.x, start.y, end.x, end.y, path.poses);
```

Comment:
Do an `if (smooth_corners_)` log warning that the edges (print IDs) couldn't make a curve between them